### ISSUE TITLE
BinarySearchQuery Bug

### ISSUE THREAD

**Issue by CamAllenSmith** [jstedfast/MailKit#967]
- OS: Windows 10
 - MailKit Version 2.4.1

**Describe the bug**
When combining two TextSearchQueries in an "And" BinarySearchQuery, the text search ignores colon.

For Example:

SearchQuery.SubjectContains("PACK:") 
     => result subjects = ["Pack: 123", "PACK: 123"]

SearchQuery.SubjectContains("PACK:").And(SearchQuery.ToContains("csmith"))
     => result subjects = ["Pack: 123", "PACK: 123", "Packing Test", ... ]

**To Reproduce**
1. Send yourself two emails, one with subject "Pack: 123" and one with subject "Packing Test"
2. Run the following code, adding your own client connection/authentication code and replacing ToContains("csmith") with your email. 

NOTES:
1.) You will need to add "using System.Linq" for the "Select" methods
2.) I have tried switching the order of the "And" query, as well as using the other overload (SearchQuery.And(left,right)). They all produce the same scenario.

```csharp
var client = new MailKit.Net.Imap.ImapClient();
client.Connect(...);
client.Authenticate(...);

// 2 results, all contain "Pack:" (with colon)
var query1 = SearchQuery.SubjectContains("Pack:");
var subjects1 = 
inbox.Search(query1)
 .Select(id => inbox.GetMessage(id))
 .Select(msg => msg.Subject).ToList();

// 25 results, all containing "pack" with/without colon
var query2 = SearchQuery.SubjectContains("Pack:").And(SearchQuery.ToContains("csmith"));
var subjects2 = 
inbox.Search(query2)
.Select(id => inbox.GetMessage(id))
.Select(msg => msg.Subject).ToList();            
```

**Expected behavior**
Only 2 results returned on the "And" SearchQuery. Refining a search with "And" should't add more results. 

**Screenshots**

subjects1 (2 results):
![image](https://user-images.githubusercontent.com/59180245/71378440-c8f3d800-2595-11ea-9ef1-9abb4bb15cbe.png)

subjects2 (25 results):
![image](https://user-images.githubusercontent.com/59180245/71378885-613e8c80-2597-11ea-88a4-29f84c1ab705.png)

**Comment by jstedfast** [comment]
Are you sure this isn't a bug with your IMAP server? Please get a [protocol log](https://github.com/jstedfast/MailKit/blob/master/FAQ.md#ProtocolLog) so I can validate that MailKit is sending the correct search query to your server.

The problem with showing me the results of your Visual Studio debug session is that it is completely irrelevant to diagnosing the bug. Either MailKit is sending the wrong search string to the server (unlikely) or the IMAP server's SEARCH implementation is broken (far more likely).

The only way to tell is to examine a protocol log.

**Comment by jstedfast** [comment]
Since MailKit does not quote the "Pack:" in the SEARCH command sent to the IMAP server, that *may* be the issue but that is technically a bug in the IMAP server because the syntax for the SEARCH command in [rfc3501](https://tools.ietf.org/html/rfc3501) is as follows:

```
search          = "SEARCH" [SP "CHARSET" SP astring] 1*(SP search-key)
                    ; CHARSET argument to MUST be registered with IANA

search-key      = "ALL" / "ANSWERED" / "BCC" SP astring /
                  "BEFORE" SP date / "BODY" SP astring /
                  "CC" SP astring / "DELETED" / "FLAGGED" /
                  "FROM" SP astring / "KEYWORD" SP flag-keyword /
                  "NEW" / "OLD" / "ON" SP date / "RECENT" / "SEEN" /
                  "SINCE" SP date / "SUBJECT" SP astring /
                  "TEXT" SP astring / "TO" SP astring /
                  "UNANSWERED" / "UNDELETED" / "UNFLAGGED" /
                  "UNKEYWORD" SP flag-keyword / "UNSEEN" /
                    ; Above this line were in [IMAP2]
                  "DRAFT" / "HEADER" SP header-fld-name SP astring /
                  "LARGER" SP number / "NOT" SP search-key /
                  "OR" SP search-key SP search-key /
                  "SENTBEFORE" SP date / "SENTON" SP date /
                  "SENTSINCE" SP date / "SMALLER" SP number /
                  "UID" SP sequence-set / "UNDRAFT" / sequence-set /
                  "(" search-key *(SP search-key) ")"
```

According to the above syntax, the `"Pack:"` string needs to be sent as an `astring` token which is defined as:

```
astring         = 1*ASTRING-CHAR / string

ASTRING-CHAR   = ATOM-CHAR / resp-specials

ATOM-CHAR       = <any CHAR except atom-specials>

atom-specials   = "(" / ")" / "{" / SP / CTL / list-wildcards /
                  quoted-specials / resp-specials

list-wildcards  = "%" / "*"

quoted-specials = DQUOTE / "\"

resp-specials   = "]"
```

Does the literal string `Pack:` meet the `1*ASTRING-CHAR` criteria? Yes, it does. Therefor it does not need to be quoted or sent as an IMAP literal token which are the other 2 ways of sending the query string.

For completeness, I'll paste the definition of `string`:

```
string          = quoted / literal

quoted          = DQUOTE *QUOTED-CHAR DQUOTE

QUOTED-CHAR     = <any TEXT-CHAR except quoted-specials> /
                  "\" quoted-specials

quoted-specials = DQUOTE / "\"

literal         = "{" number "}" CRLF *CHAR8
                    ; Number represents the number of CHAR8s
```

To verify that my diagnosis is correct, you can check the results of the following code snippet:

```csharp
var subjects3 = ((ImapFolder) inbox).Search ("SUBJECT \"Pack:\" TO \"csmith\"")
    .Select(id => inbox.GetMessage(id))
    .Select(msg => msg.Subject).ToList(); 
```

**Comment by jstedfast** [comment]
When you paste your protocol log, make sure to include the greeting and also to scrub your AUTHENTICATE command.

**Comment by jstedfast** [comment]
FWIW, try testing this for yourself using a free GMail account. It works exactly like it should.

This just further confirms that this is not a bug in MailKit.

**Comment by CamAllenSmith** [comment]
I appreciate all the feedback and apologize for the delayed response, I was out of the office for the holiday. I believe you that it's working perfectly for gmail but unfortunately we require this for an office365/outlook account. 

I ran your code for subjects3 and still got 25 matches. I had to slightly modify it to the following (including UniqueIds because the SearchResults object isn't enumerable)

`
  var subjects3 = ((ImapFolder)inbox).Search("SUBJECT \"Pack:\" TO \"csmith\"").UniqueIds
                                    .Select(id => inbox.GetMessage(id))
                                    .Select(msg => msg.Subject).ToList();
`

![image](https://user-images.githubusercontent.com/59180245/71481349-87e50900-27cb-11ea-93a4-206f9f9445ee.png)


Below is the log output, I ran this on an account with an empty inbox to make it much shorter, (my main account created a log with ~88k lines) so let me know if you need the full one.

Connected to imaps://outlook.office365.com:993/
S: * OK The Microsoft Exchange IMAP4 service is ready. [...]
C: A00000000 CAPABILITY
S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
S: A00000000 OK CAPABILITY completed.
C: A00000001 AUTHENTICATE PLAIN ...
S: A00000001 OK AUTHENTICATE completed.
C: A00000002 CAPABILITY
S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION BACKENDAUTHENTICATE CHILDREN IDLE NAMESPACE LITERAL+
S: A00000002 OK CAPABILITY completed.
C: A00000003 NAMESPACE
S: * NAMESPACE (("" "/")) NIL NIL
S: A00000003 OK NAMESPACE completed.
C: A00000004 LIST "" "INBOX"
S: * LIST (\Marked \HasNoChildren) "/" INBOX
S: A00000004 OK LIST completed.
C: A00000005 EXAMINE INBOX
S: * 1 EXISTS
S: * 1 RECENT
S: * FLAGS (\Seen \Answered \Flagged \Deleted \Draft $MDNSent)
S: * OK [PERMANENTFLAGS ()] Permanent flags
S: * OK [UNSEEN 1] Is the first unseen message
S: * OK [UIDVALIDITY 14] UIDVALIDITY value
S: * OK [UIDNEXT 41] The next unique identifier value
S: A00000005 OK [READ-ONLY] EXAMINE completed.
C: A00000006 UID SEARCH SUBJECT Pack:
S: * SEARCH
S: A00000006 OK SEARCH completed.
C: A00000007 UID SEARCH SUBJECT Pack: TO csmith
S: * SEARCH
S: A00000007 OK SEARCH completed.
C: A00000008 UID SEARCH SUBJECT "Pack:" TO "csmith"
S: * SEARCH
S: A00000008 OK SEARCH completed.
C: A00000009 LOGOUT
S: * BYE Microsoft Exchange Server IMAP4 server signing off.
S: A00000009 OK LOGOUT completed.

**Comment by jstedfast** [comment]
At this point it's pretty clearly a bug in Office365 and if my suggestion about quoting the strings didn't work around the bug, then I don't know what else to suggest other than to do your searching locally after downloading all of the messages.

The reason I suggested trying to quote the strings by using the raw search expression API was to see if *maybe* the problem was just the IMAP server's SEARCH expression parser was just broken and somehow getting confused by the unquoted `:` character. That doesn't seem to be the case, though, since you got the exact same results even after quoting the strings.

This means that the bug is in the IMAP server's *SEARCH implementation* and not their parser which means there's no way to wrk around it using syntax hacks.

**Comment by jstedfast** [comment]
> I don't know what else to suggest other than to do your searching locally after downloading all of the messages.

Let me rephrase this... I would probably still recommend using the IMAP SEARCH feature, just know that you'll have to re-run the equivalent search filter locally to weed out the incorrect matches.

This would be a better strategy than downloading *all* of the messages in the folder and doing a local search on them (unless, of course, you are downloading all of the messages anyway).

**Comment by CamAllenSmith** [comment]
Thank you Jeffrey. We will submit this bug to Microsoft and cross our fingers. I appreciate all the help.
